File: resonanse_common/src/models/accounts.rs
use sqlx::postgres::PgRow;
use sqlx::{Error, FromRow, Row};

#[derive(Debug)]
pub struct UserData {
    pub first_name: String,
    pub last_name: String,
    pub city: Option<String>,
    pub description: Option<String>,
    pub headline: Option<String>,
    pub goals: Option<String>,
    pub interests: Option<String>,
    pub language: Option<String>,
    pub age: Option<i16>,
    pub education: Option<String>,
    pub hobby: Option<String>,
    pub music: Option<String>,
    pub sport: Option<String>,
    pub books: Option<String>,
    pub food: Option<String>,
    pub worldview: Option<String>,
}

#[derive(Debug)]
pub struct UserTgData {
    pub tg_username: Option<String>,
    pub tg_user_id: Option<i64>,
}

#[derive(Debug)]
pub struct UserContactData {
    pub email: Option<String>,
    pub phone: Option<String>,
    pub telegram: UserTgData,
    pub instagram: Option<String>,
}

#[derive(Debug)]
pub struct AuthData {
    pub password_hash: Option<String>,
}

#[repr(i32)]
#[derive(Clone, Copy, Debug, sqlx::Type)]
pub enum ResoAccountType {
    Standard = 0,
    Bad = 1,
    Banned = 2,
    Premium = 3,
}

#[derive(Debug)]
pub struct BaseAccount {
    pub id: i64,
    pub username: Option<String>,
    pub user_data: UserData,
    pub contact_data: UserContactData,
    pub auth_data: AuthData,
    pub user_type: ResoAccountType,
}

impl FromRow<'_, PgRow> for BaseAccount {
    fn from_row(row: &PgRow) -> Result<Self, Error> {
        Ok(BaseAccount {
            id: row.try_get("id")?,
            username: row.try_get("username")?,
            user_data: UserData {
                first_name: row.try_get("first_name")?,
                last_name: row.try_get("last_name")?,
                city: row.try_get("city")?,
                description: row.try_get("description")?,
                headline: row.try_get("headline")?,
                goals: row.try_get("goals")?,
                interests: row.try_get("interests")?,
                language: row.try_get("language")?,
                age: row.try_get("age")?,
                education: row.try_get("education")?,
                hobby: row.try_get("hobby")?,
                music: row.try_get("music")?,
                sport: row.try_get("sport")?,
                books: row.try_get("books")?,
                food: row.try_get("food")?,
                worldview: row.try_get("worldview")?,
            },
            contact_data: UserContactData {
                email: row.try_get("email")?,
                phone: row.try_get("phone")?,
                telegram: UserTgData {
                    tg_username: row.try_get("tg_username")?,
                    tg_user_id: row.try_get("tg_user_id")?,
                },
                instagram: row.try_get("instagram")?,
            },
            auth_data: AuthData {
                password_hash: row.try_get("password_hash")?,
            },
            user_type: row.try_get("user_type")?,
        })
    }
}

File: backend_py/services/accounts.py
from typing import Optional

from fastapi import HTTPException, Depends, status
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend_py.app import app, OpenApiTags
from backend_py.db import UserAccountDB, get_db


# Модель Pydantic для запроса обновления аккаунта
class UpdateAccountRequest(BaseModel):
    first_name: str
    last_name: str
    city: str
    about: str
    headline: Optional[str]
    goals: Optional[str]
    interests: Optional[str]
    language: Optional[str]
    age: Optional[int]
    education: Optional[str]
    hobby: Optional[str]
    music: Optional[str]
    sport: Optional[str]
    books: Optional[str]
    food: Optional[str]
    worldview: Optional[str]
    alcohol: Optional[str]
    email: Optional[str]
    phone: Optional[str]
    tg_username: Optional[str]
    tg_user_id: Optional[int]
    instagram: Optional[str]


# Модель Pydantic для ответа с информацией об аккаунте
class AccountInfo(BaseModel):
    id: int
    username: str
    first_name: str
    last_name: str
    city: str
    about: str
    user_type: int

    class Config:
        orm_mode = True
        from_attributes = True


# Обновление информации об аккаунте
@app.put("/api/accounts/{account_id}", response_model=AccountInfo, tags=[OpenApiTags.ACCOUNTS])
async def update_account(account_id: int, updated_info: UpdateAccountRequest, db: Session = Depends(get_db)):
    account = db.query(UserAccountDB).filter(UserAccountDB.id == account_id).first()
    if account is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Account not found")

    # Обновляем только те поля, которые предоставлены в запросе
    for field, value in updated_info.dict().items():
        setattr(account, field, value)

    db.commit()
    db.refresh(account)
    return AccountInfo.from_orm(account)


# Получение информации об аккаунте
@app.get("/api/accounts/{account_id}", response_model=AccountInfo, tags=[OpenApiTags.ACCOUNTS])
async def get_account(account_id: int, db: Session = Depends(get_db)):
    account = db.query(UserAccountDB).filter(UserAccountDB.id == account_id).first()
    if account is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Account not found")
    return AccountInfo.model_validate(account)


# Удаление аккаунта
@app.delete("/api/accounts/{account_id}", tags=[OpenApiTags.ACCOUNTS])
async def delete_account(account_id: int, db: Session = Depends(get_db)):
    account = db.query(UserAccountDB).filter(UserAccountDB.id == account_id).first()
    if account is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Account not found")

    db.delete(account)
    db.commit()

    return {"message": "Account deleted successfully"}
